File: String Hashing/String_Hashing/rabin_karp.cpp
//Rabin Karp Algorithm
//Language Used: C++
//Rabin Karp algorithm to search for a pttern in a string in linear time complexity
//Input Format: First line contains text, second line contains pattern to be searched
//Output Format: Output contains only one line i.e., starting index of found string else not found
//Sample Input:  abbababab
//               abab
//Sample Output: Pattern found starting from 3 index
//author:sarthakeddy

#include<bits/stdc++.h>
using namespace std;
#define ll long long

void search(string pattern, string text, ll prime)
{
	ll pat_len=pattern.length();
	ll text_len=text.length();
	ll hash_text=0, hash_pat=0, h=1;
	ll i,j;

	for(i=0; i<pat_len-1; i++)
		h= (h*256) % prime;

	for(i=0;i<pat_len;i++)
	{
		hash_pat= (256*hash_pat + pattern[i]) %prime;
		hash_text= (256*hash_text + text[i]) %prime;
	}

	for(i=0;i<=(text_len-pat_len);i++)
	{
		if(hash_text == hash_pat)
		{
			for(j=0;j<pat_len;j++)
			{
				if(text[i+j] != pattern[j])
					break;
			}
			if(j==pat_len)
				cout<<"Pattern found starting from "<<i<<" index\n";
		}

		if(i<(text_len-pat_len))
		{
			hash_text=(256* (hash_text - text[i]*h) + text[i+pat_len]) %prime;
			if(hash_text < 0)
				hash_text+=prime;
		}
	}
}


int main()
{
	string text,pattern;
	cout<<"Enter the text\n";
	cin>>text;
	cout<<"Enter the pattern to be searched\n";
	cin>>pattern;
	search(pattern,text,101);
}

File: rabin_karp.java
import java.util.Scanner;
public class rabin_karp  
{ 
    public final static int d = 256; 
    static void search(String pat, String txt, int q) 
    { 
        int M = pat.length(); 
        int N = txt.length(); 
        int i, j; 
        int p = 0; // hash value for pattern 
        int t = 0; // hash value for txt 
        int h = 1; 
        for (i = 0; i < M-1; i++) 
            h = (h*d)%q; 
        for (i = 0; i < M; i++) 
        { 
            p = (d*p + pat.charAt(i))%q; 
            t = (d*t + txt.charAt(i))%q; 
        }
        for (i = 0; i <= N - M; i++) 
        { 
            if ( p == t ) 
            { 
                for (j = 0; j < M; j++) 
                { 
                    if (txt.charAt(i+j) != pat.charAt(j)) 
                        break; 
                } 
       
                if (j == M) 
                    System.out.println("Pattern found at index " + i); 
            } 
            if ( i < N-M ) 
            { 
                t = (d*(t - txt.charAt(i)*h) + txt.charAt(i+M))%q; 
                if (t < 0) 
                t = (t + q); 
            } 
        } 
    } 
    public void main() 
    {
        Scanner sc =new Scanner(System.in);
        System.out.println("Enter the text:");
        String txt =sc.next(); 
        System.out.println("Enter the pattern:");
        String pat =sc.next(); 
        int q = 101; // A prime number 
        search(pat, txt, q); 
    } 
} 